src/docfx/lib/log/SourceInfo{T}.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+namespace Microsoft.Docs.Build
+{
+    [DebuggerDisplay("{Value}")]
+    public sealed class SourceInfo<T> : SourceInfo
+    {
+        public readonly T Value;
+
+        internal override object GetValue() => Value;
+
+        public SourceInfo(T value, SourceInfo source = null)
+            : base(source?.File, source?.Line ?? 0, source?.Column ?? 0, source?.EndLine ?? 0, source?.EndColumn ?? 0)
+        {
+            Value = value;
+        }
+
+        public SourceInfo<T> WithValue(T value)
+        {
+            if (value == default)
+                return null;
```

Comment:
line info will be lost if you return `null` here. If `Value` needs permutation, we can make `Value` settable.

Reply:
yes, it was not easy to make it settable in the previous PR, will update it here